20. Valid Parentheses.md

olsen-blue opened the pull request:
・問題
https://leetcode.com/problems/valid-parentheses/description/

・次やる問題
https://leetcode.com/problems/reverse-linked-list/description/

olsen-blue committed:
20. Valid Parentheses.md

oda commented on 20. Valid Parentheses.md:
```diff
@@ -0,0 +1,73 @@
+        for char in s:
+            if char in '({[':           #"({["でも、同じらしい。
+                stack.append(char)
+            else:
+                if not stack: return False
+                if char == ')' and stack.pop(-1) != '(': return False
+                if char == '}' and stack.pop(-1) != '{': return False
+                if char == ']' and stack.pop(-1) != '[': return False
```
私は、ここで一回 continue にして、下、全部インデントを下げます。

stack.pop が条件式の中にあるのは少し怖くて、これ、char に )}] でないものが来ると、pop されないわけですね。
最終的に、最後の not stack でひっかかるはずなのでいいんですが、これって考えにあった動作でしたでしょうか。
副作用のある条件式は意外な動作をすることがあります。

olsen-blue replied:
> char に )}] でないものが来ると、pop されないわけですね。
こちらですが、「 )}] でないもの」というのは、「)}]」以外の入力(例：a, abc, 1, 12)を想定されていますでしょうか？これは、想定できていませんでした。(){}[]のみが入力としてあるという想定でした。

結果的に"({[a})"のような入力であれば、Falseを返せてOK（<-これが最後のnot stockでひっかかるということですかね）ですが、例えば、"({[a]})"のような入力だと、aの処理をスルーするので、False返すべきところをTrueになってしまいますね...

oda replied:
(私の上のコメント少々混乱しています。pop されないから、その文字はなかったことになりますね。)

はい、問題文から a などは来ません。ただ、。現実には、おそらく、このコードがプロダクションで使われていたところ、ある日、まったく違う事情でここにおかしな文字が流れるようになって、予期せぬ動作をして、原因探しの旅に出てここに行き着くことになります。そのときにどのような動作をこのコードはしていますか。みたいな想像をしています。
https://discord.com/channels/1084280443945353267/1316770883729100810/1322841071301955684

そういうときに何が起きるコードかを念頭に置いて、それでよいかどうかを考えておきましょうということです。

olsen-blue replied:
なるほどです。問題を解くことばかりに注意が向いて視野が狭くなっており、失礼いたしました。実際の現場では、自分の書いたコードが、将来的に起き得る意図しない状況下でもしっかりと機能するように配慮しなければならない、ということですね。
ソフトウェアエンジニアリングとはプログラミングを時間積分したような物である、と聞いた記憶があるのですが、まさに今回仰っていただいたような事でしょうか。初めて実感を持って理解できた気がします。ありがとうございます。
https://zenn.dev/voicy/articles/72840463657d2a

olsen-blue replied:
Step4として、意図しない入力にも対応できるよう、コードを修正いたしました。取り急ぎご共有までにと思います。
こっちの書き方だと、if＆elseではなく、if+continueの書き方がしっくりときますね。
https://github.com/olsen-blue/Arai60/pull/6/files

olsen-blue committed:
Update 20. Valid Parentheses.md

olsen-blue committed:
Step4：意図しない入力への対応を追加

odaさんにいただいたアドバイスに沿って、修正にトライ。
https://github.com/olsen-blue/Arai60/pull/6#discussion_r1904177090

t0hsumi reviewed (commented):
いいと思います。

`open_to_close = {'(': ')'}`みたいな感じでdictで置いておくと、例えば追加問題として、あらたに'<', '>'も考慮してくださいとなったときに、このdictの書き換えのみですみループないはいじらなくてすむなと思いました。

hroc135 commented on 20. Valid Parentheses.md:
```diff
@@ -0,0 +1,106 @@
+##Step1
```
マークダウンで描きたい場合、##の後に半角スペースを入れるとその行の文字が大きくなります
#1個が最大で数が増える小さくなっていきます

olsen-blue replied:
Github初心者なので助かります。他にも変なところがあれば、ぜひよろしくお願いいたします。

olsen-blue commented:
似たような[問題](https://leetcode.com/problems/asteroid-collision/)で、自分にとってより直感的なリファクタリングができて嬉しかったので、記録します。
beforeのコードは、[解説動画](https://www.youtube.com/watch?v=LN7KjRszjk4)を参考にしていたのですが、astというint変数１つで、小惑星(ast)の爆発有無管理&ループ脱出管理をしていました。ちょっと１変数の荷が重すぎるのと、forで取り出したものを0で書き換えることに認知負荷の高さを感じました。
そこで、分離されたものが好みなので分離(爆発有無はbool変数で別管理＆ループ脱出はbreakで即座に行う)をしたところ、afterのコードでは、よりスムーズに書けるようになり嬉しいです。
（同じ問題を復習で何度も解いていますが、いまだに前向きな発見（<- もっと良い体の動かし方を見つけてしまった...!）があったりで、それが楽しいです。）

(cf: リファクタリング以外の感想だと、高校物理の運動量をイメージすると見通しが良い点と、for で ast を固定し今までの軌跡を振り返って積み上げたstackに対して処理をするのは monotonic stack のアルゴリズムでやっていることと非常に近い点、が嬉しい気づきでした。)

- before
```py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for ast in asteroids:
            while stack and stack[-1] > 0 and ast < 0:
                total = stack[-1] + ast
                if total == 0:
                    stack.pop()
                    ast = 0
                if total < 0:
                    stack.pop()
                if total > 0:
                    ast = 0
            if ast != 0:
                stack.append(ast)
        return stack
```

- after
```py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for ast in asteroids:
            is_alive = True
            while stack and stack[-1] > 0 and ast < 0:
                total = stack[-1] + ast
                if total == 0:
                    stack.pop()
                    is_alive = False
                    break
                if total > 0:
                    is_alive = False
                    break
                if total < 0:
                    stack.pop()
            if is_alive:
                stack.append(ast)
        return stack
```

oda commented:
フラグを立てるならば while-else を使うか、is_alive を返す関数に切り出すという手があります。
「左行きスタック(確定)」と「右行きスタック(バッファー)」に分けるというのもありかと思います。

olsen-blue commented:
odaさんありがとうございます。確認遅くなりすみません。気になった解法で書いてみました。

- while-else:
whileループ + elseの後続処理、を早期 break で切り上げられるのが良いと思いました。
```py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for ast in asteroids:
            while stack and stack[-1] > 0 and ast < 0:
                total = stack[-1] + ast
                if total == 0:
                    stack.pop()
                    break
                if total < 0:
                    stack.pop()
                if total > 0:
                    break
            else:
                stack.append(ast)
        return stack
```

- 関数切り出し:
何度もやる処理は関数化すべきですね。
関数の return で ast の運命が即座にスパッと決定される感じが良いと感じました。
(stackの中を削除する機能も付属している旨は docstring などに書きたいと思いました。)
```py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        def is_survived(ast, stack) -> bool:
            """
            Overview:
                forで取り出した小惑星 ast が生存するかどうかを返す 
            Note:
                衝突発生によるスタック内の要素(追加済みの ast )削除機能もあり
            """
            if not stack:
                return True
            while stack and stack[-1] > 0 and ast < 0:
                total = stack[-1] + ast
                if total == 0:
                    stack.pop()
                    return False
                if total < 0:
                    stack.pop()
                if total > 0:
                    return False
            return True
        
        stack = []
        for ast in asteroids:
            if is_survived(ast, stack):
                stack.append(ast)
        return stack
```
